accord-core/src/main/java/accord/coordinate/Barrier.java

```diff
@@ -0,0 +1,309 @@
+                check);
+        return check;
+    }
+
+    // Hold result of looking for a transaction to act as a barrier for an Epoch
+    static class BarrierTxn
+    {
+        @Nonnull
+        public final TxnId txnId;
+        @Nonnull
+        public final Timestamp executeAt;
+        @Nonnull
+        public final Status status;
+        @Nonnull
+        public final Key key;
+        public BarrierTxn(@Nonnull TxnId txnId, @Nonnull Timestamp executeAt, @Nonnull Status status, Key key)
+        {
+            this.txnId = txnId;
+            this.executeAt = executeAt;
+            this.status = status;
+            this.key = key;
+        }
+
+        public BarrierTxn max(BarrierTxn other)
+        {
+            if (other == null)
+                return this;
+            return status.compareTo(other.status) >= 0 ? this : other;
+        }
+    }
+
+    /*
+     * Check for an existing transaction that is either already Applied (dependencies were applied)
+     * or PreApplied (outcome fixed, dependencies not yet applied).
+     *
+     * For Applied we can return success immediately with the executeAt epoch. For PreApplied we can add
+     * a listener for when it transitions to Applied and then return success.
+     */
+    class ExistingTransactionCheck extends AsyncResults.AbstractResult<BarrierTxn> implements MapReduceConsume<SafeCommandStore, BarrierTxn>
+    {
+        @Override
+        public BarrierTxn apply(SafeCommandStore safeStore)
+        {
+            BarrierTxn found = safeStore.mapReduceWithTerminate(
+                    seekables,
+                    safeStore.ranges().allSince(minEpoch),
+                    // TODO review this with Benedict, but I don't believe an exclusive sync point works here
+                    TestKind.Any,
+                    TestTimestamp.EXECUTES_AFTER,
+                    TxnId.minForEpoch(minEpoch),
+                    TestDep.ANY_DEPS,
+                    null,
+                    // TODO is PreApplied right? It's the first OutcomeKnown state, but why not Committed?
+                    Status.PreApplied,
+                    Status.Applied,
+                    (keyOrRange, txnId, executeAt, status, barrierTxn) -> {
+                        if (keyOrRange.domain() == Domain.Key)
+                            return new BarrierTxn(txnId, executeAt, status, keyOrRange.asKey());
+                        return null;
+                    },
+                    null,
+                    // Take the first one we find, and call it good enough to wait on
+                    barrierTxn -> barrierTxn != null);
+            // It's not applied so add a listener to find out when it is applied
+            if (found != null && !found.status.equals(Status.Applied))
+            {
+                safeStore.commandStore().execute(contextFor(found.txnId), safeStoreWithTxn -> {
+                    // TODO handle null return from get due to truncation
```

Comment:
is this TODO still relevant?

Reply:
I don't think it is, it won't return null right? It will return something non-null for the transaction id saying it's not known.